server/observable.go

```diff
@@ -504,12 +527,50 @@ func (o *Observable) updateStore() {
 	}
 }
 
+func (o *Observable) shouldSample() bool {
+	if o.sampleFrequency <= 0 {
+		return false
+	}
+
+	if o.sampleFrequency >= 100 {
+		return true
+	}
+
+	if mrand.Int31n(100) <= o.sampleFrequency {
+		return true
+	}
+
+	return false
+}
+
+func (o *Observable) sampleAck(sseq, dseq uint64) {
+	if !o.shouldSample() {
+		return
+	}
+
+	e := &ObservableSampleEvent{
+		MsgSet:     o.mset.Name(),
+		Observable: o.name,
+		ObsSeq:     dseq,
+		Delay:      int64(time.Now().UnixNano()) - o.pending[sseq],
+	}
+
+	j, err := json.MarshalIndent(e, "", "  ")
+	if err != nil {
+		return
+	}
+
+	target := JetStreamObservableSamplePre + "." + e.MsgSet + "." + e.Observable
```

Comment:
This makes a bit of garbage, would create it once on obs creation or use strings.Join (IIRC)

Reply:
good call.